File: src/DataProvider/Player.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace WGApiDataProvider
{
    public class Player
    {
        public readonly int ID;
        public readonly string Name;

        private readonly int ClanID;
        private static Clan EmptyClan = new Clan(0, String.Empty, String.Empty, 0, 0, 0, 0, 0, 0, 0, new WGApi.ClanEmblems());
        [JsonIgnore]
        public Clan Clan
        {
            get
            {
                if (ClanID == 0)
                    return EmptyClan;
                if (WGApiDataProvider.Instance._Clans.TryGetValue(ClanID, out Clan result))
                    return result;
                return EmptyClan;
            }
        }

        private static DataProvider DataProvider;
        public static void SetDataProvider(DataProvider provider) => DataProvider = provider;

        [JsonIgnore]
        public int ThreeMoeCount
        {
            get
            {
                if (DataProvider._PlayersMarks.TryGetValue(ID, out var result))
                    return result.Count;
                return 0;
            }
        }

        public readonly int BattleCount;
        public readonly decimal WinRatio;
        public readonly double Wn8;
        public double MoeRating
        {
            get
            {
                if (DataProvider._PlayersMarks.TryGetValue(ID, out var result))
                    return result.Values.Sum(x => x.Tank.MoeValue);
                return 0;
            }
        }
        public readonly int WgRating;
        public readonly string ClientLanguage;
        public readonly DateTime LastBattle;
        public readonly DateTime LastLogout;
        public readonly DateTime AccountCreated;
        public readonly DateTime LastWgUpdate;
        public readonly DateTime LastChecked;

        public Player(
            int id,
            string name,
            int clanID,
            int threeMoeCount,
            int battleCount,
            decimal winRatio,
            double wn8,
            double moeRating,
            int wgRating,
            string clientLanguage,
            int lastBattle,
            int lastLogout,
            int accountCreated,
            int lastWgUpdate,
            int lastChecked
        ) : this(id, name, clanID, threeMoeCount, battleCount, winRatio, wn8, moeRating, wgRating, clientLanguage,
                 WGApi.EpochDateTime.FromEpoch(lastBattle), WGApi.EpochDateTime.FromEpoch(lastLogout),
                 WGApi.EpochDateTime.FromEpoch(accountCreated), WGApi.EpochDateTime.FromEpoch(lastWgUpdate),
                 WGApi.EpochDateTime.FromEpoch(lastChecked)) { }

        [JsonConstructor]
        public Player(
            int id,
            string name,
            int clanID,
            int threeMoeCount,
            int battleCount,
            decimal winRatio,
            double wn8,
            double moeRating,
            int wgRating,
            string clientLanguage,
            DateTime lastBattle,
            DateTime lastLogout,
            DateTime accountCreated,
            DateTime lastWgUpdate,
            DateTime lastChecked
        )
        {
            ID = id;
            Name = name;
            ClanID = clanID;
            //ThreeMoeCount = threeMoeCount;
            BattleCount = battleCount;
            WinRatio = winRatio;
            Wn8 = wn8;
            WgRating = wgRating;
            ClientLanguage = clientLanguage;
            LastBattle = lastBattle;
            LastLogout = lastLogout;
            AccountCreated = accountCreated;
            LastWgUpdate = lastWgUpdate;
            LastChecked = lastChecked;
        }

        public Player(WGApi.PlayerInfo player)
        {
            ID = player.AccountID;
            Name = player.Nick;
            ClanID = player.ClanID ?? 0;
            //ThreeMoeCount = -1;
            BattleCount = player.Statistics.Random.Battles;
            WinRatio = player.Statistics.Random.Victories / Math.Max(1m, player.Statistics.Random.Battles);
            Wn8 = -1;
            WgRating = player.WGRating;
            ClientLanguage = player.ClientLanguage;
            LastBattle = player.LastBattle;
            LastLogout = player.LastLogout;
            AccountCreated = player.AccountCreated;
            LastWgUpdate = player.UpdatedAt;
            LastChecked = DateTime.Now;
        }
    }
}

File: src/3MoeSite/View/Player.fs
[<AutoOpen>]
module _3MoeSite.Views.Player
open Giraffe
open Giraffe.GiraffeViewEngine
open _3MoeSite.Model
open System
open System.Linq
open WGApiDataProvider
open _3MoeSite.Views
open _3MoeSite

let data = WGApiDataProvider.Instance

let playerPage (id : int) (params : TableParams) =
    let (success, player) = data._Players.TryGetValue id

    let statBlockColoredBackground (value : StatValueObject) (title : string) (colorClass : string)= 
        div [ _class (sprintf "statValueBox darkBorder %s" colorClass) ] [
                div [ _class "statValueBoxDiv" ] [ encodedText (printStatValueObject value) ]
                encodedText title
            ]

    let statBlock (value : StatValueObject) (title : string) = 
        div [ _class "statValueBox" ] [
                div [ _class "statValueBoxDiv" ] [ encodedText (printStatValueObject value) ]
                encodedText title
            ]        

    match success with
    | false -> errorBlock (sprintf "Could not find the player with Wargaming ID '%i'" id)
    | true ->
        [
            div [ _class "playerInfoBlock"] [
                div [ _class "clearfix" ] [
                    div [ _class "playerInfoDiv"] [
                        div [ _class "playerName" ] [
                            encodedText player.Name
                            img [ _src (Links.clientLanguageFlag player.ClientLanguage) ]
                        ]
                        div [ _class "moeCount"] [
                        img [ _src "/img/3stars.png" 
                              _class "image3moe" ]
                        encodedText (System.String.Format("{0:N0}", player.ThreeMoeCount))
                        ]
                        div [ _class "playerLinkDiv"] [
                            linkWithImage (Links.playerProfileWG player.ID) "WG Profile " "http://eu.wargaming.net/favicon.ico"
                            linkWithImage (Links.playerProfileWotlabs player.Name) "Wotlabs" "http://wotlabs.net/images/favicon.png"
                        ]
                    ]
                    div [ _class (sprintf "clanInfoDiv %s" (if player.Clan.ID = 0 then "hidden" else "" )) ] [
                        img [ _src player.Clan.Emblems.x32.Portal  
                              _class "clanIcon" ]
                        span [ _class "clanTag" ] [ 
                                a [ _href (Links.clanPage player.Clan.ID)
                                    _target "blank" ] [
                                        encodedText player.Clan.Tag
                                ]
                        ]
                        div [ _class "clanLinkDiv" ] [
                            linkWithImage (Links.clanProfileWG player.Clan.ID) "WG Profile" "http://eu.wargaming.net/favicon.ico"
                            linkWithImage (Links.clanProfileWotlabs player.Clan.Tag) "Wotlabs" "http://wotlabs.net/images/favicon.png"
                        ]
                    ]
                ]
                div [ _class "valueBoxes clearfix"] [
                    statBlockColoredBackground (StatValueObject.F player.Wn8) "WN8" (getWn8CssClass player.Wn8)
                    statBlockColoredBackground (StatValueObject.D player.WinRatio) "Winratio" (getWinratioCssClass player.WinRatio)
                    statBlock (StatValueObject.I player.BattleCount) "Battles"
                    statBlock (StatValueObject.F player.MoeRating) "MoE Rating"
                    statBlock (StatValueObject.I player.WgRating) "WG Rating"
                ]
                div [ _class "valueBoxes clearfix"] [
                    dateBlock player.LastBattle "Last Battle"
                    dateBlock player.LastLogout "Last Logout"
                    dateBlock player.LastChecked "Last checked"
                    dateBlock player.AccountCreated "Account created"
                    dateBlock player.LastWgUpdate "Last update (WG)"
                ]
            ]

            (match data._PlayersMarks.TryGetValue player.ID with
            | false, _ -> div [ _class "noMarksDiv" ][ 
                            encodedText "This player has no 3 Marks of Excellence on any tank."                
                          ]
            | true, marks ->
                marks.Values |> Seq.toArray |> customTable ([
                    createColumn "Contour" "contour" (fun m -> Img m.Tank.Icons.Contour)
                    createCustomColumn "Tank"         "tank"  (fun m -> S m.Tank.Name)
                        (fun m -> (a [ _href (Links.tankPage m.Tank.ID) ] [
                                        encodedText m.Tank.Name ]))
                    createColumn "First detected"   "firstDetection"    (fun m -> T m.FirstDetected)
                    createColumn "3 MoE"        "moe"   (fun m -> TableCellObject.I m.Tank.ThreeMoeCount)
                    createColumn "MoE Value"    "moev"  (fun m -> TableCellObject.D m.Tank.MoeValue)
                    createCustomColumn "Tier"         "tier"  (fun m -> TableCellObject.I m.Tank.Tier)
                        (fun m -> (linkWithImage (Links.tierPage m.Tank.Tier) (string m.Tank.Tier) ""))
                    createCustomColumn "Nation"       "nat"   (fun m -> E m.Tank.Nation)
                        (fun m -> (linkWithImage (Links.nationPage ((int)m.Tank.Nation)) (string m.Tank.Nation) (Links.nationFlag ((string m.Tank.Nation).ToLower()))))
                    createCustomColumn "Type"         "type"  (fun m -> E m.Tank.VehicleType)
                        (fun m -> (linkWithImage (Links.vehicleTypePage (int m.Tank.VehicleType)) (string m.Tank.VehicleType) ""))
                    ] : Mark Column List) params)
        ] |> layout player.Name


let playersTable params =
    let tableTemplate = customTable ([
        createCustomColumn "Name"       "name"      (fun p -> S p.Name)
            (fun p -> a [ _href (Links.playerPage p.ID) ] [ encodedText p.Name ])
        createCustomColumn "Clan"       "clan"      (fun p -> S (string p.Clan.Tag))
            (fun p -> clanLinkWithIcon p.Clan)
        createColumn "3 MoE"            "moe"       (fun p -> TableCellObject.I p.ThreeMoeCount)
        createColumn "Battles"          "battles"   (fun p -> TableCellObject.I p.BattleCount)
        createColumn "Win ratio"        "wr"        (fun p -> Perc p.WinRatio)
        createColumn "WN8"              "wn8"       (fun p -> TableCellObject.D p.Wn8)
        createColumn "MoE Rating"       "moer"      (fun p -> TableCellObject.D p.MoeRating)
        createColumn "WG Rating"        "wgr"       (fun p -> TableCellObject.I p.WgRating)
        createColumn "Client language"  "lang"      (fun p -> Flag p.ClientLanguage)
        createColumn "Last battle"      "lbatt"     (fun p -> T p.LastBattle)
        createColumn "Last logout"      "llog"      (fun p -> T p.LastLogout)
        createColumn "Account created"  "created"   (fun p -> T p.AccountCreated)
        createColumn "Last update (WG)" "lupd"      (fun p -> T p.LastWgUpdate)
        createColumn "Last checked"     "lch"       (fun p -> T p.LastChecked)
    ] : Player Column list) params

    [
        headlineBlock (System.String.Format("{0:N0} Players", data.Players.Length))
        data.Players |> tableTemplate
    ] |> layout "Players"


let playerHandler (id : int) (params : TableParams) =
    htmlView (playerPage id params)